### ISSUE TITLE
Remove order merge logic happening on every request

### ISSUE THREAD

**Issue by mtomov** [solidusio/solidus#1116]
Hello,

[set_current_order](https://github.com/solidusio/solidus/blob/master/core/lib/spree/core/controller_helpers/order.rb#L60) is a method, which is called on [every request](https://github.com/solidusio/solidus/blob/master/core/lib/spree/core/controller_helpers/order.rb#L11) and results in 3 sql queries for logged in users:

``` sql
  Spree::Order Load (0.7ms)  SELECT  "spree_orders".* FROM "spree_orders" WHERE "spree_orders"."completed_at" IS NULL AND "spree_orders"."currency" = $1 AND "spree_orders"."guest_token" = $2 AND "spree_orders"."store_id" = $3 AND "spree_orders"."user_id" = $4 LIMIT 1  [["currency", "USD"], ["guest_token", "6YHZ-S_2WypaPxxky4fHVA"], ["store_id", 2], ["user_id", 1]]
  Spree::Order Load (0.4ms)  SELECT  "spree_orders".* FROM "spree_orders" WHERE "spree_orders"."user_id" = $1 AND "spree_orders"."frontend_viewable" = $2 AND "spree_orders"."store_id" = 2  ORDER BY "spree_orders"."created_at" DESC LIMIT 1  [["user_id", 1], ["frontend_viewable", "t"]]
  Spree::Order Load (0.3ms)  SELECT "spree_orders".* FROM "spree_orders" WHERE "spree_orders"."user_id" = $1 AND "spree_orders"."completed_at" IS NULL AND (id != 9)  [["user_id", 1]]


```

![spree orders - 2 queries](https://cloud.githubusercontent.com/assets/1651750/15068768/720a75ec-1371-11e6-9f22-6668cb8eef88.png)

The method performs an order merge operation if there's more than one incomplete order for the user. 

Unless I miss a use case, this _correct behaviour_ is have this operation be performed only after a user login - exactly in [UserSessionsController#create](https://github.com/solidusio/solidus_auth_devise/blob/master/lib/controllers/frontend/spree/user_sessions_controller.rb#L15). 

It is currently incorrectly executed there in a `before_action` instead of an `after_action` with the inclusion of the [`Order`](https://github.com/spree/spree_auth_devise/blob/master/lib/controllers/frontend/spree/user_sessions_controller.rb#L6) module.
## 

<del>
I was wondering if anyone would have a recommendation on how to cleanly move the order merge logic in the create session operation, as stores could be using different gems for authentication.

- One way is through registering after login hooks, like the `self.update_hooks = Set.new` on [Order.rb](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/order.rb#L147), but it seems like too much of a change.
</del>

---

Edit, after having another look, it appears that you have already come up with a brilliant solution to that:

[config/initializers/warden.rb](https://github.com/solidusio/solidus_auth_devise/blob/master/config/initializers/warden.rb#L2)

So, do you think it will be a good idea to tuck the Order#merge logic in there as well? I'm only not sure how to get the `current_order` from warden, but there must be a way.

@mvz - I'm pinging you as you were active on this area of the code recently.

Thank you!

**Comment by cbrunsdon** [comment]
This actually looks like a pretty reasonable approach to me. We will need to ensure that we have something friendly for non-auth-devise usage as well but I think its pretty low hanging fruit to remove every round trips from every request. Would be a good target for someone during the hack days next week, too...

**Comment by jhawthorn** [comment]
Fixed by #2185
